pkg/parsing/lex/run_state.go

```diff
@@ -0,0 +1,173 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package lex
+
+import (
+	"fmt"
+	"strings"
+	"unicode/utf8"
+
+	"github.com/tricksterproxy/trickster/pkg/parsing/token"
+)
+
+// RunState contains all the information about a particular run
+type RunState struct {
+	Input        string            // the string being scanned
+	InputLowered string            // the lowercase version of the string being scanned
+	InputWidth   int               // width of the input (so we don't have to keep calling len)
+	Pos          int               // current position in the input
+	Start        int               // start position of this Token
+	Width        int               // width of last rune read from input
+	Tokens       chan *token.Token // channel of scanned items
+	ParenDepth   int               // nesting depth of ( ) exprs
+}
+
+// Next returns the next rune in the input.
+func (rs *RunState) Next() rune {
+	if int(rs.Pos) >= rs.InputWidth {
+		rs.Width = 0
+		return EOF
+	}
+	r, w := utf8.DecodeRuneInString(rs.InputLowered[rs.Pos:])
+	rs.Width = w
+	rs.Pos += rs.Width
+	return r
+}
+
+// Peek returns but does not consume the next rune in the input.
+func (rs *RunState) Peek() rune {
+	r := rs.Next()
+	rs.Backup()
+	return r
+}
+
+// Backup steps back one rune. Can only be called once per call of next.
+func (rs *RunState) Backup() {
+	rs.Pos -= rs.Width
+}
+
+// Emit passes an Token back to the client.
+func (rs *RunState) Emit(t token.Typ) {
+	rs.Tokens <- &token.Token{Typ: t, Pos: rs.Start, Val: rs.InputLowered[rs.Start:rs.Pos]}
+	rs.Start = rs.Pos
+}
+
+// EmitToken passes a pre-built an Token back to the client. Returning nil allows the
+// function output to be used as a value when needd to consolidate EmitToken and the
+// likely subsequent return nil
+func (rs *RunState) EmitToken(i *token.Token) StateFn {
+	rs.Tokens <- i
+	rs.Start = rs.Pos
+	return nil
+}
+
+// Ignore skips over the pending input before this point.
+func (rs *RunState) Ignore() {
+	rs.Start = rs.Pos
+}
+
+// Accept consumes the next rune if it's from the valid set.
+func (rs *RunState) Accept(valid string) bool {
+	if strings.ContainsRune(valid, rs.Next()) {
+		return true
+	}
+	rs.Backup()
+	return false
+}
+
+// AcceptRun consumes a run of runes from the valid set.
+func (rs *RunState) AcceptRun(valid string) bool {
+	var b bool
+	for strings.ContainsRune(valid, rs.Next()) {
+		b = true
+	}
+	rs.Backup()
+	return b
+}
+
+// // Drain drains the output so the lexing goroutine will exit.
+// // Called by the parser, not in the lexing goroutine.
+// func (rs *RunState) Drain() {
```

Comment:
Ded code?

Reply:
We're unsure if the Drain func will be needed yet. Once ClickHouse and InfluxDB implement the parser, we might need to run a drain in the event of a lexing error, so that orphan channels don't lie around indefinitely.